pkg/os/os_windows.go

```diff
@@ -0,0 +1,120 @@
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package os
+
+import (
+	"os"
+	"strings"
+
+	"golang.org/x/sys/windows"
+)
+
+// openPath takes a path, opens it, and returns the resulting handle.
+// It works for both file and directory paths.
+//
+// We are not able to use builtin Go functionality for opening a directory path:
+// - os.Open on a directory returns a os.File where Fd() is a search handle from FindFirstFile.
+// - syscall.Open does not provide a way to specify FILE_FLAG_BACKUP_SEMANTICS, which is needed to
+//   open a directory.
+// We could use os.Open if the path is a file, but it's easier to just use the same code for both.
+// Therefore, we call windows.CreateFile directly.
+func openPath(path string) (windows.Handle, error) {
+	u16, err := windows.UTF16PtrFromString(path)
+	if err != nil {
+		return 0, err
+	}
+	h, err := windows.CreateFile(
+		u16,
+		0,
+		windows.FILE_SHARE_READ|windows.FILE_SHARE_WRITE|windows.FILE_SHARE_DELETE,
+		nil,
+		windows.OPEN_EXISTING,
+		windows.FILE_FLAG_BACKUP_SEMANTICS, // Needed to open a directory handle.
+		0)
+	if err != nil {
+		return 0, &os.PathError{
+			Op:   "CreateFile",
+			Path: path,
+			Err:  err,
+		}
+	}
+	return h, nil
+}
+
+// GetFinalPathNameByHandle flags.
+//nolint:golint
+const (
+	cFILE_NAME_OPENED = 0x8
+
+	cVOLUME_NAME_DOS  = 0x0
+	cVOLUME_NAME_GUID = 0x1
+)
+
+// getFinalPathNameByHandle facilitates calling the Windows API GetFinalPathNameByHandle
+// with the given handle and flags. It transparently takes care of creating a buffer of the
+// correct size for the call.
+func getFinalPathNameByHandle(h windows.Handle, flags uint32) (string, error) {
+	n, err := windows.GetFinalPathNameByHandle(h, nil, 0, flags)
+	if err != nil {
+		return "", err
+	}
+	b := make([]uint16, n)
+	_, err = windows.GetFinalPathNameByHandle(h, &b[0], n, flags)
+	if err != nil {
+		return "", err
+	}
+	return windows.UTF16ToString(b), nil
+}
+
+// resolvePath implements path resolution for Windows. It attempts to return the "real" path to the
+// file or directory represented by the given path.
+// The resolution works by using the Windows API GetFinalPathNameByHandle, which takes a handle and
+// returns the final path to that file.
+func resolvePath(path string) (string, error) {
+	h, err := openPath(path)
+	if err != nil {
+		return "", err
+	}
+	rPath, err := getFinalPathNameByHandle(h, cFILE_NAME_OPENED|cVOLUME_NAME_DOS)
+	if err == windows.ERROR_PATH_NOT_FOUND {
+		// ERROR_PATH_NOT_FOUND is returned from the VOLUME_NAME_DOS query if the path is a
+		// volume GUID path to a disk, and the disk does not have an assigned drive letter.
+		// In this case, just get the volume GUID path instead.
+		rPath, err = getFinalPathNameByHandle(h, cFILE_NAME_OPENED|cVOLUME_NAME_GUID)
```

Comment:
What is the difference between the volume GUID path that we used in the `openPath` and the volume GUID path that is returned here?

Reply:
The string passed to `openPath` could be any sort of path string, it's not necessarily a volume GUID path.

Reply:
So if the `resolvePath` is called with a path like `\\?\Volume{GUID}` then the first call to `getFinalPathNameByHandle` would fail with `ERROR_PATH_NOT_FOUND` and the second call would return the same path again?

Reply:
That's correct, but that will hold true for any path resolution function. If you call it with an already fully resolved path, it should return the same thing as the input.

It might be helpful to consider the case of a symlink `lnk -> \\?\Volume{abcd}\`. In that case calling `resolvePath("lnk")` should resolve the path and return `\\?\Volume{abcd}\`.

Reply:
Ah it makes sense now. I think it would be helpful to slightly change that comment to make it say `ERROR_PATH_NOT_FOUND is returned from the VOLUME_NAME_DOS query if the _final_ path is a volume GUID path to a disk`.

Reply:
Thanks, that's good feedback. :) I'm actually currently working on rewriting the comments here to be much more useful. I also ended up needing to switch the order so we do VOLUME_NAME_GUID first, due to another issue in `EvalSymlinks` (and wanting to ensure the output of `resolvePath` can be passed to `EvalSymlinks` safely).